### ISSUE TITLE
Allow some way of specifying 'isoformat' as the 'date_format' parameter in 'to_csv' methods

### ISSUE THREAD

**Issue by nugend** [pandas-dev/pandas#27328]
#### Code Sample, a copy-pastable example if possible

```
>>> import pandas as pd
>>> mydf = pd.DataFrame.from_dict(dict(dt=[pd.Timestamp('2000-01-01'),pd.Timestamp('2000-01-01 00:00:00.000000001')]))
>>> print(mydf.to_csv(index=False))
dt
2000-01-01 00:00:00.000000000
2000-01-01 00:00:00.000000001
>>> print(mydf.to_csv(index=False,date_format='%Y-%m-%dT%H:%M:%S.%f'))
dt
2000-01-01T00:00:00.000000
2000-01-01T00:00:00.000000
```
#### Problem description

ISO-8601 specifies that the 'T' separator is optional in contexts where there is no ambiguity with other 8601 date or time formats. I'm trying to make some CSV files and am trying to be very strict with what I emit because I don't know how they'll be consumed.

It'd be *really* nice to be able to preserve full precision in the data but still have that separator character present and as far as I am aware, there is no Python Datetime format string which displays nanoseconds.

The shortest alternative I can come up with for this is the following (and I am not 100% sure this will work correctly):

```
tmp_df = mydf
dt_cols = mydf.select_dtypes(include='datetime64').columns
tmp_df[dt_cols] = tmp_df[dt_cols].applymap(lambda x:str(x.to_datetime64()))
tmp_df.to_csv(index=False)
```

#### Expected Output

```
dt
2000-01-01T00:00:00.000000000
2000-01-01T00:00:00.000000001
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.8.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-327.36.3.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.2
pytest: 3.6.4
pip: 19.1.1
setuptools: 41.0.1
Cython: None
numpy: 1.16.4
scipy: None
pyarrow: 0.13.0
xarray: None
IPython: 7.5.0
sphinx: None
patsy: None
dateutil: 2.8.0
pytz: 2019.1
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.2.4
openpyxl: 2.5.14
xlrd: 1.1.0
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: 1.2.19
pymysql: None
psycopg2: None
jinja2: 2.10.1
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

**Comment by mroeschke** [comment]
We send the output to `datetime.strftime`, so as you alluded to, this is a limitation of the standard library and unlikely to be fixed directly in pandas.

The alternative is to cast to string before saving to csv. This could be valuable to add to our documentation.

```
In [8]: print(mydf.dt.astype(str).to_csv(index=False))
2000-01-01 00:00:00.000000000
2000-01-01 00:00:00.000000001
```

**Comment by nugend** [comment]
I don't believe that's completely correct. AFAICT, the datetime string is emitted from this function: https://github.com/pandas-dev/pandas/blob/54c0d5a7aad1c09f9c50d8fe218187f5e1e6d1e8/pandas/_libs/tslib.pyx#L220-L300

You can see on line 269 that the datetime string is manually constructed there. Is there any possibility of either detecting a specific date_format with an otherwise unused control character (Ruby uses `%N`, for example), or maybe a `iso8601sep` binary flag or something? I know the former is nasty and the latter involves updating a bunch of callsites to pass a boolean through, but this doesn't need to touch `datetime.strftime`

**Comment by khalludi** [comment]
Using the template from mroeschke, you can add the letter 'T' pretty easily by replacing the space character.

```
>>> mydf.dt.astype(str).apply(lambda x: x.replace(' ', 'T'))
0    2000-01-01T00:00:00.000000000
1    2000-01-01T00:00:00.000000001
Name: dt, dtype: object
```

I'm still new here, but I think the above workaround solves the issue of the optional T without having to change the codebase. Therefore, the issue can be closed.

Please correct me if I'm wrong! Linking @mroeschke for feedback.

**Comment by nugend** [comment]
I'll have to try and remember what we did. I *think* it might have been something similar.

**Comment by pcoccoli** [comment]
Why does `pandas.DataFrame.to_json` accept `'iso'` as `date_format` but `to_csv` doesn't?  `to_string` doesn't accept a `date_format` arg at all.  It all seems so arbitrary.

**Comment by tommyjcarpenter** [comment]
Totally agree with @pcoccoli ; `to_any_media_type()` should have the same args - I see no semantic difference in timestamps that `to_json` would support `iso` but `to_csv` would not - just doesn't make much sense.

**Comment by Simon-Bru** [comment]
This is really annoying everytime having to come up with the date format, I would like to submit my 🆙

**Comment by andrewmwilson** [comment]
Python 3.12 adds a `%:z` format string which solves this issue.
